services/src/main/java/org/keycloak/services/clienttype/impl/DefaultClientType.java

```diff
@@ -81,47 +84,84 @@ public <T> T getDefaultValue(String optionName, Class<T> optionType) {
+        ClientRepresentation currentRep = ModelToRepresentation.toRepresentation(currentClient, session);
+        validateClientRequest(newClient, currentRep);
+    }
+
+    protected void validateClientRequest(ClientRepresentation newClient, ClientRepresentation currentClient) throws ClientTypeException {
+        List<String> validationErrors = clientType.getConfig().entrySet().stream()
+                .filter(property -> clientPropertyHasInvalidChangeRequested(currentClient, newClient, property.getKey(), property.getValue()))
+                .map(Map.Entry::getKey)
+                .collect(Collectors.toList());
+
+        if (validationErrors.size() > 0) {
+            throw new ClientTypeException(
+                    "Cannot change property of client as it is not allowed by the specified client type.",
+                    validationErrors.toArray());
+        }
+    }
+
+    protected ClientRepresentation augmentClient(ClientRepresentation client) {
+        clientType.getConfig().entrySet()
+                .forEach(property -> setClientProperty(client, property.getKey(), property.getValue()));
+        return client;
+    }
+
+    private boolean clientPropertyHasInvalidChangeRequested(
+            ClientRepresentation oldClient,
+            ClientRepresentation newClient,
+            String propertyName,
+            ClientTypeRepresentation.PropertyConfig propertyConfig) {
+        Object newClientProperty = getClientProperty(newClient, propertyName);
+        Object oldClientProperty = getClientProperty(oldClient, propertyName);
+
+        return (
+                    // Validate that non-applicable client properties were not changed.
+                    !propertyConfig.getApplicable() &&
+                    !Objects.isNull(newClientProperty) &&
+                    !Objects.equals(oldClientProperty, newClientProperty)
+                ) || (
+                    // Validate that applicable read-only client properties were not changed.
+                    propertyConfig.getApplicable() &&
+                    propertyConfig.getReadOnly() &&
+                    !Objects.isNull(newClientProperty) &&
+                    !Objects.equals(oldClientProperty, newClientProperty)
+                );
+    }
+
+    private void setClientProperty(ClientRepresentation client,
+                               String propertyName,
+                               ClientTypeRepresentation.PropertyConfig propertyConfig) {
+
+        if (!propertyConfig.getApplicable() || propertyConfig.getDefaultValue() == null) {
+            return;
+        }
+
+        if (clientRepresentationProperties.containsKey(propertyName)) {
+            // Java property on client representation
+            Method setter = clientRepresentationProperties.get(propertyName).getWriteMethod();
+            try {
+                setter.invoke(client, propertyConfig.getDefaultValue());
```

Comment:
Looks good. 

FYI. from the long term perspective, I can think about avoid using reflection and instead rely on something like:

```
Map<String, BiConsumer<ClientModel, Object>> properties;

properties.put("standardFlowEnabled", (clientModel, standardFlowModel) -> client.setStandardFlowEnabled(standardFlowEnabled));
```

And then in `setClientProperty` something like:

```
BiConsumer invoker = properties.get(propertyName);
invoker.accept(client, propertyConfig.getDefaultValue());
```

This can hopefully work for "compound" properties as wel (like clientScopes or protocolaMappers). But that is something for the future...

One question here is, if we need to set properties at all during client create/update as if properties are provided by clientType itself, we can rely on the properties returned by the clientType (done in `TypeAwareClientModelDelegate`) instead of rely on the properties returned by the client?

But those are all open questions for future discussions, rather than something specific to this PR :-)

Reply:
Agreed! Having a consistent way of augmenting the client model / representation would be a good idea.